### PR TITLE
Improves error messages, adds NULL checks and replaces localtime with localtime_r

### LINKED ISSUE #41: Potential NULL pointer dereference
In the following part of the code, we get the username that is associated with a particular user ID:

https://github.com/zakkak/turnin/blob/d2b800db86f2ece3d160ea7e9466e52738b79d38/src/turnin.c#L468

The code correctly copies the string to a local variable, as according to the man page, there's no guarantee that this value will continue to live there long term (it's in the static section, and is overwritten by subsequent calls).

However, according to the man page, there's no guarantee that the struct fields that are pointers to `char` arrays will not be `NULL`. Although in practice this always worked, I think it's much safer to introduce a `NULL` check to avoid a dereference.

### PR DESCRIPTION
**Pull request by alx01011**
Improves error messages on `seteuid`. `NULL` checks are now performed before dereferencing `pw_name` .
Also, `localtime` is replaced by `localtime_r`, although overwrite risk is minimal. It's return value is also checked before dereferencing.

Should fix #41 #42 #43 .

### CODE REVIEW

File: src/turnin.c [hunk 1]
```diff
@@ -208,10 +208,18 @@ void check_assignment(char* s){
 
 char * timestamp(time_t clock) {
 	char *b = (char *) malloc(16);
-	struct tm *t = localtime(&clock);
+	struct tm t;
+	struct tm *res = localtime_r(&clock, &t);
```
Comment (daknob) [inline src/turnin.c:212]: I found this in the `man` page:

> According to POSIX.1-2004, localtime() is required to behave as though [tzset](https://linux.die.net/man/3/tzset)(3) was called, while localtime_r() does not have this requirement. For portable code [tzset](https://linux.die.net/man/3/tzset)(3) should be called before localtime_r().

Can you add this call please?

File: src/turnin.c [hunk 2]
```diff
@@ -465,6 +473,13 @@ void setup(char *arg) {
 		        "        Please report this issue to the system administrators\n", user_uid);
 		exit(1);
 	}
+
+	if (!pwd->pw_name) {
+		fprintf(stderr, 
+					  "turnin: Cannot lookup user_name (uid %d)\n"
+						" 			 Please report this issue to the system administrators\n", user_uid);
```
Comment (daknob) [inline src/turnin.c:480]: I think these are not aligned with the rest of the file.

File: src/turnin.c [hunk 3]
```diff
@@ -1364,14 +1380,26 @@ void checkdue() {
 	diff_days = (int)diff_time / 86400; // in days
 
 	// calculate penalty
-	tm_curr = localtime(&curr_time);
-	penalty += (tm_curr->tm_wday == 0 || tm_curr->tm_wday == 6) ? weekendpenalty : daypenalty;
+	res = localtime_r(&curr_time, &tm_curr);
+	if (!res) {
+		fprintf(stderr,	
+				"turnin: Cannot create timestamp\n"
```
Comment (daknob) [inline src/turnin.c:1386]: Maybe make this line different between the 3 cases that can result in it, so you can figure out where exactly was the problem in the code.

**Review by daknob — CHANGES_REQUESTED** [review]

File: src/turnin.c (hunk 1 above)
Comment (alx01011) [reply src/turnin.c:212]: Will do, thanks for the pointer.

File: src/turnin.c (hunk 2 above)
Comment (alx01011) [reply src/turnin.c:480]: Yes you are right, it looked fine on my end. Is it ok if I include a `.clang-format` to avoid same issues in the future?

Comment (daknob) [reply src/turnin.c:480]: @zakkak What do you think?

Comment (zakkak) [reply src/turnin.c:480]: +1 ideally it should be accompanied by an [`.editorconfig`](https://editorconfig.org/#example-file) file

**Comment by alx01011** [comment]
@daknob @zakkak If you have time, please review 07eb70429a83bd08d410ee9857dde5243de103dd, e2ab5e7c09c64bef69c73b4401611ab5c4ec4430 and 0649ebd833698d3d96068523bd2b8d4fa64ac7bf.

I have included `LLVM`'s `.clang-format` style, it's slightly different than the indentation used before. 
I have also included a naive `.editorconfig`. 

Both files can be tweaked and improved to match the projects specifications.

**Review by zakkak — CHANGES_REQUESTED** [review]
Hi @alx01011 ,

Can you please apply the following to the clang format?

```diff
diff --git a/src/.clang-format b/src/.clang-format
index 48b2c67..1b71d18 100644
--- a/src/.clang-format
+++ b/src/.clang-format
@@ -4,10 +4,10 @@ Language:        Cpp
 AccessModifierOffset: -2
 AlignAfterOpenBracket: Align
 AlignArrayOfStructures: None
-AlignConsecutiveMacros: None
-AlignConsecutiveAssignments: None
-AlignConsecutiveBitFields: None
-AlignConsecutiveDeclarations: None
+AlignConsecutiveMacros: Consecutive
+AlignConsecutiveAssignments: Consecutive
+AlignConsecutiveBitFields: Consecutive
+AlignConsecutiveDeclarations: Consecutive
 AlignEscapedNewlines: Right
 AlignOperands:   Align
 AlignTrailingComments: true
@@ -15,7 +15,7 @@ AllowAllArgumentsOnNextLine: true
 AllowAllParametersOfDeclarationOnNextLine: true
 AllowShortEnumsOnASingleLine: true
 AllowShortBlocksOnASingleLine: Never
-AllowShortCaseLabelsOnASingleLine: false
+AllowShortCaseLabelsOnASingleLine: true
 AllowShortFunctionsOnASingleLine: All
 AllowShortLambdasOnASingleLine: All
 AllowShortIfStatementsOnASingleLine: Never
@@ -57,7 +57,7 @@ BreakConstructorInitializersBeforeComma: false
 BreakConstructorInitializers: BeforeColon
 BreakAfterJavaFieldAnnotations: false
 BreakStringLiterals: true
-ColumnLimit:     80
+ColumnLimit:     120
 CommentPragmas:  '^ IWYU pragma:'
 QualifierAlignment: Leave
 CompactNamespaces: false
@@ -179,9 +179,9 @@ StatementAttributeLikeMacros:
 StatementMacros:
   - Q_UNUSED
   - QT_REQUIRE_VERSION
-TabWidth:        8
+TabWidth:        2
 UseCRLF:         false
-UseTab:          Never
+UseTab:          AlignWithSpaces
 WhitespaceSensitiveMacros:
   - STRINGIZE
   - PP_STRINGIZE
@@ -189,4 +189,3 @@ WhitespaceSensitiveMacros:
   - NS_SWIFT_NAME
   - CF_SWIFT_NAME
 ...
-

```

This will bring it closer to the original style. 

Furthermore, I would prefer any indentation changes to be in a separate commit (i.e. don't apply the new clang-format rules and commit them with other changes).

Thanks

**Comment by zakkak** [comment]
Oh, one more thing. Please move `.clang-format` and `.editorconfig` to the root directory of the project.

**Comment by alx01011** [comment]
@zakkak Can you review 0b1a0b52a75999f5925907bc687603fc042aed7e? 0b1a0b52a75999f5925907bc687603fc042aed7e contains the indentation changes on `turnin.c`.

**Comment by zakkak** [comment]
Superseded by #50 and #51

**Pull request closed without merging**